src/Data/Tree.purs

```diff
@@ -9,8 +13,35 @@ import Data.Foldable (Foldable, foldr, foldl, foldMap)
 import Data.Traversable (Traversable, traverse, sequence)
 import Data.NonEmpty (NonEmpty(), (:|))
 
+-- | An abstract tree type
 data Tree a = Tree a (Array (Tree a))
 
+-- | A view which is used to construct and destruct the tree. This type is a
+-- | pattern functor/signature, where `f` represents the recursive structure.
+data View f a = View a (Array (f a))
+
+-- | Pattern match on a tree. Example:
+-- |
+-- | ````purescript
+-- | case out t of
+-- |   View a ts -> ...
+-- | ````
+out :: forall a. Tree a -> View Tree a
+out (Tree a ts) = View a ts
+
+-- | Construct a tree from a tree pattern/view. Example:
+-- |
+-- | ````purescript
+-- | tree :: Tree Int
+-- | tree = into $ View 0 []
+-- | ````
+into :: forall a. View Tree a -> Tree a
+into (View a ts) = Tree a ts
+
+-- | Map a tree across a natural transformation.
+transView :: forall f g a. (forall a. f a -> g a) -> View f a -> View g a
```

Comment:
Interesting stuff!

Is this going to give a warning in the future? https://github.com/purescript/purescript/issues/1465